src/v/security/acl_store.h

```diff
@@ -10,71 +10,115 @@
  */
 #pragma once
 #include "security/acl.h"
-#include "utils/absl_sstring_hash.h"
+#include "security/acl_entry_set.h"
 
 #include <absl/container/btree_map.h>
-#include <absl/container/flat_hash_map.h>
-#include <absl/container/node_hash_set.h>
+
+#include <ranges>
 
 namespace security {
 
 /*
- * A collection of ACL entries.
+ * Container for ACLs.
  */
-class acl_entry_set {
+class acl_matches;
+class acl_store {
 public:
-    using container_type = absl::node_hash_set<acl_entry>;
-    using role_cache_type
-      = absl::flat_hash_map<ss::sstring, ssize_t, sstring_hash, sstring_eq>;
-    using const_iterator = container_type::const_iterator;
-    using const_reference = std::reference_wrapper<const acl_entry>;
-
-    acl_entry_set() noexcept = default;
-    acl_entry_set(acl_entry_set&&) noexcept = default;
-    acl_entry_set& operator=(acl_entry_set&&) noexcept = default;
-    acl_entry_set(const acl_entry_set&) = delete;
-    acl_entry_set& operator=(const acl_entry_set&) = delete;
-    ~acl_entry_set() noexcept = default;
-
-    void insert(acl_entry entry);
-    void erase(container_type::const_iterator it) {
-        remove_if_role(it->principal());
-        _entries.erase(it);
-    }
-    void remove_if_role(const acl_principal_base& p);
-
-    template<typename Predicate>
-    void erase_if(Predicate pred) {
-        absl::erase_if(_entries, pred);
-    }
+    acl_store() = default;
+    acl_store(acl_store&&) noexcept = default;
+    acl_store& operator=(acl_store&&) noexcept = default;
+    acl_store(const acl_store&) = delete;
+    acl_store& operator=(const acl_store&) = delete;
+    ~acl_store() noexcept = default;
 
-    void rehash() {
-        _entries.rehash(0);
-        _role_cache.rehash(0);
+    void add_bindings(const std::vector<acl_binding>& bindings) {
+        for (auto& binding : bindings) {
+            auto& entries = _acls[binding.pattern()];
+            entries.insert(binding.entry());
+            entries.rehash();
```

Comment:
rehash after the loop? seems like rehashing after each insert could be an expensive thing?

Reply:
yeh nbd, just drive by noticing things. maybe i had added this long ago?

Reply:
Ya, I assume it was from the OG implementation. I looked at it again and I think we'd need a container or something to keep track of which entries matched, so maybe not much to be gained anyway